tests/System.Text.Primitives.Tests/EncodingPerfTestsUsingTextEncoder.cs

```diff
@@ -0,0 +1,316 @@
+                        if (!utf8.TryEncode(characters, span, out int consumed, out int written)) { throw new Exception(); }
+            }
+        }
+
+        [Benchmark(InnerIterationCount = InnerCount)]
+        [MemberData(nameof(GetEncodingPerformanceTestData))]
+        public void EncodeFromUtf32toUtf8UsingTextEncoder(int length, int minCodePoint, int maxCodePoint, SpecialTestCases special = SpecialTestCases.None)
+        {
+            string inputString = GenerateStringData(length, minCodePoint, maxCodePoint, special);
+            char[] characters = inputString.AsSpan().ToArray();
+            Encoding utf32 = Encoding.UTF32;
+            int utf32Length = utf32.GetByteCount(characters);
+            var utf32Buffer = new byte[utf32Length];
+            utf32.GetBytes(characters, 0, characters.Length, utf32Buffer, 0);
+            Span<byte> utf32ByteSpan = new Span<byte>(utf32Buffer);
+            ReadOnlySpan<uint> utf32Span = utf32ByteSpan.NonPortableCast<byte, uint>();
+
+            TextEncoder utf8 = TextEncoder.Utf8;
+            Assert.True(utf8.TryComputeEncodedBytes(utf32Span, out int encodedBytes));
+            var utf8Buffer = new byte[encodedBytes];
+            var span = new Span<byte>(utf8Buffer);
+
+            foreach (var iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                    for (int i = 0; i < Benchmark.InnerIterationCount; i++)
+                        if (!utf8.TryEncode(utf32Span, span, out int consumed, out int written)) { throw new Exception(); }
+            }
+        }
+
+        [Benchmark(InnerIterationCount = InnerCount)]
+        [MemberData(nameof(GetEncodingPerformanceTestData))]
+        public void EncodeFromUtf8toUtf16UsingTextEncoder(int length, int minCodePoint, int maxCodePoint, SpecialTestCases special = SpecialTestCases.None)
+        {
+            string inputString = GenerateStringData(length, minCodePoint, maxCodePoint, special);
+            char[] characters = inputString.AsSpan().ToArray();
+            Encoding utf8Encoder = Encoding.UTF8;
+            int utf8Length = utf8Encoder.GetByteCount(characters);
+            var utf8TempBuffer = new byte[utf8Length];
+            utf8Encoder.GetBytes(characters, 0, characters.Length, utf8TempBuffer, 0);
+            Span<byte> utf8Span = new Span<byte>(utf8TempBuffer);
+
+            TextEncoder utf16 = TextEncoder.Utf16;
+            Assert.True(utf16.TryComputeEncodedBytes(utf8Span, out int encodedBytes));
+            var utf16Buffer = new byte[encodedBytes];
+            var span = new Span<byte>(utf16Buffer);
+
+            foreach (var iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                    for (int i = 0; i < Benchmark.InnerIterationCount; i++)
+                        if (!utf16.TryEncode(utf8Span, span, out int consumed, out int written)) { throw new Exception(); }
```

Comment:
Note, the biggest cause of the difference is a result of the below comment since we aren't comparing the same thing.